### PR TITLE
Refactor usage of Nats on type-level

### PR DESCRIPTION
**Pull request by hovanja2011**

### CODE REVIEW

File: tests/Main.hs [hunk 1]
```diff
@@ -41,10 +41,10 @@ main = do
     -- Arithmetic circuit
     specArithmeticCircuit
 
-    -- Arithmetization
+     -- Arithmetization
```
Comment (vlasin) [inline tests/Main.hs:44]: Why extra space

File: tests/Main.hs [hunk 2]
```diff
@@ -41,10 +41,10 @@ main = do
...
     specArithmetization
 
-    -- Protocols
+     -- Protocols
```
Comment (vlasin) [inline tests/Main.hs:47]: Extra space, again

File: src/ZkFold/Symbolic/Data/Helpers.hs [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+---
+unsafeDiv :: forall a b. Dict (Div a b * b ~ a)
+unsafeDiv = unsafeAxiom
+
+withUnsafeDiv :: forall a b {r}. ( (Div a b * b ~ a) => r) -> r
+withUnsafeDiv = withDict (unsafeDiv @a @b)
+
+---
+divNN :: forall a. Dict (Div a a ~ a)
```
Comment (vks4git) [inline src/ZkFold/Symbolic/Data/Helpers.hs:54]: Where is this constraint used? Doesn't `Div a a ~ a` imply that `a ~ 1`?

**Review by vks4git — DISMISSED** [review]

File: src/ZkFold/Symbolic/Data/Helpers.hs [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+---
+divisibleDiv :: forall a b. (Mod a b ~ 0) :- (Div a b * b ~ a)
+divisibleDiv = Sub unsafeAxiom
+
+withDivisibleDiv :: forall a b {r}. Mod a b ~ 0 => ((Div a b * b ~ a) => r) -> r
+withDivisibleDiv = withDict (divisibleDiv @a @b)
+
+---
+gcdn8 :: forall a. Dict (Gcd a 8 ~ 8)
```
Comment (TurtlePU) [inline src/ZkFold/Symbolic/Data/Helpers.hs:54]: This is definitely not true

File: src/ZkFold/Symbolic/Data/ByteString.hs [hunk 5]
```diff
@@ -232,24 +211,16 @@ instance (Symbolic c, KnownNat n) => BoolType (ByteString n c) where
...
 -- (the list can contain an arbitrary number of words).
 -- We can only impose some restrictions on @n@ and @m@.
 --
-
-instance
-  ( Symbolic c
-  , (Div k m) * m ~ k
-  ) => Concat (ByteString m c) (ByteString k c) where
-    concat bs = (ByteString . packed) $ V.unsafeConcat @(Div k m) ( Haskell.map (\(ByteString bits) -> unpacked bits) bs)
+concat :: forall m k n c. (Symbolic c, k * m ~ n) => Vector k (ByteString m c) -> ByteString n c
```
Comment (TurtlePU) [inline src/ZkFold/Symbolic/Data/ByteString.hs:222]: Can the return type be just `ByteString (k * m) c`?

File: src/ZkFold/Symbolic/Data/ByteString.hs [hunk 6]
```diff
@@ -232,24 +211,16 @@ instance (Symbolic c, KnownNat n) => BoolType (ByteString n c) where
 -- 4. @wordSize@ divides @n@.
 --
 
-instance
-  ( Symbolic c
-  , KnownNat wordSize
-  , (Div n wordSize) * wordSize ~ n
-  ) => ToWords (ByteString n c) (ByteString wordSize c) where
-    toWords (ByteString bits) = Haskell.map (ByteString . packed) $ V.fromVector . V.chunks @(Div n wordSize) @wordSize $ unpacked bits
+toWords :: forall n wordSize c. (Symbolic c, KnownNat wordSize, Div n wordSize * wordSize ~ n) => ByteString n c -> Vector (Div n wordSize) (ByteString wordSize c)
```
Comment (TurtlePU) [inline src/ZkFold/Symbolic/Data/ByteString.hs:214]: It's better to avoid `Div`, like this:
```haskell
toWords :: (Symbolic c, KnownNat wordSize) => ByteString (m * wordSize) c -> Vector m (ByteString wordSize c)
```

File: src/ZkFold/Symbolic/Data/ByteString.hs [hunk 7]
```diff
@@ -69,18 +70,20 @@ deriving via (Structural (ByteString n c))
          instance (Symbolic c) => Eq (Bool c) (ByteString n c)
 
 instance
-    ( FromConstant Natural (ByteString 8 c)
-    , Concat (ByteString 8 c) (ByteString n c)
+    ( Symbolic c
+    , FromConstant Natural (ByteString 8 c)
```
Comment (TurtlePU) [inline src/ZkFold/Symbolic/Data/ByteString.hs:74]: This should be implied by `Symbolic c`

File: src/ZkFold/Symbolic/Data/ByteString.hs [hunk 8]
```diff
@@ -69,18 +70,20 @@ deriving via (Structural (ByteString n c))
...
+    , m * 8 ~ n
     ) => IsString (ByteString n c) where
     fromString = fromConstant . fromString @Bytes.ByteString
 
 instance
-    ( FromConstant Natural (ByteString 8 c)
-    , Concat (ByteString 8 c) (ByteString n c)
+    ( Symbolic c
+    , FromConstant Natural (ByteString 8 c)
```
Comment (TurtlePU) [inline src/ZkFold/Symbolic/Data/ByteString.hs:81]: This should be implied by `Symbolic c`

File: src/ZkFold/Symbolic/Data/Helpers.hs [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+import           Prelude
+
+import           ZkFold.Base.Algebra.Basic.Number
+
+
+-- type ExtensionBits inputLen = 8 * (128 - Mod inputLen 128)
+-- type ExtendedInputByteString inputLen c = ByteString (8 * inputLen + ExtensionBits inputLen) c
+
+
+with4n6 :: forall n {r}. KnownNat n => (KnownNat (4 * n + 6) => r) -> r
```
Comment (TurtlePU) [inline src/ZkFold/Symbolic/Data/Helpers.hs:17]: As this is used in Plonk code, it's better to move it there. Maybe near `PlonkupPolyExtendedLength` definition is a good place.

File: src/ZkFold/Symbolic/Data/Helpers.hs
Comment (TurtlePU) [inline src/ZkFold/Symbolic/Data/Helpers.hs:1]: This is not Symbolic-specific, and actually I think can be placed in `ZkFold.Base.Algebra.Basic.Number` after cleaning it up a little

Comment (TurtlePU) [reply src/ZkFold/Symbolic/Data/Helpers.hs:1]: (Also better to name the `Basic.Number` module a little better, something like `ZkFold.Base.Data.Type.Nat` or just `ZkFold.Base.Data.TypeNats`)

File: src/ZkFold/Symbolic/Data/UInt.hs [hunk 10]
```diff
@@ -172,7 +172,6 @@ instance
     , KnownNat n
     , KnownNat k
     , KnownRegisterSize r
-    , n <= k
```
Comment (TurtlePU) [inline src/ZkFold/Symbolic/Data/UInt.hs:175]: This is a meaningful constraint, better to leave it here

File: src/ZkFold/Symbolic/Data/Helpers.hs [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+with4n6 f = withDict (timesNat @4 @n) (withDict (plusNat @(4 * n) @6) f)
+
+with8n  :: forall n {r}. KnownNat n => (KnownNat (8 * n) => r) -> r
+with8n = withDict (timesNat @8 @n)
+
+withExtensionBits :: forall n {r}. KnownNat n => (KnownNat (8 * (128 - Mod n 128)) => r) -> r
```
Comment (TurtlePU) [inline src/ZkFold/Symbolic/Data/Helpers.hs:23]: Better to move all "domain-specific" functions closer to their domains

File: src/ZkFold/Symbolic/Data/ByteString.hs [hunk 12]
```diff
@@ -232,24 +211,16 @@ instance (Symbolic c, KnownNat n) => BoolType (ByteString n c) where
...
 -- We can only impose some restrictions on @n@ and @m@.
 --
-
-instance
-  ( Symbolic c
-  , (Div k m) * m ~ k
-  ) => Concat (ByteString m c) (ByteString k c) where
-    concat bs = (ByteString . packed) $ V.unsafeConcat @(Div k m) ( Haskell.map (\(ByteString bits) -> unpacked bits) bs)
+concat :: forall m k n c. (Symbolic c, k * m ~ n) => Vector k (ByteString m c) -> ByteString n c
+concat bs = (ByteString . packed) $ V.concat ( V.parFmap (\(ByteString bits) -> unpacked bits) bs)
```
Comment (TurtlePU) [inline src/ZkFold/Symbolic/Data/ByteString.hs:223]: Also maybe we can make this a little faster by using `packWith`

File: tests/Tests/Blake2b.hs
Comment (TurtlePU) [inline tests/Tests/Blake2b.hs:1]: Looks like a part of work on Blake2b, not on a typenats refactor, so maybe it's better to do it in a different PR

File: tests/Tests/UInt.hs [hunk 13]
```diff
@@ -64,7 +65,6 @@ specUInt'
     :: forall p n r r2n rs
     .  PrimeField (Zp p)
     => KnownNat n
-    => KnownNat (2 * n)
     => KnownRegisterSize rs
     => n <= 2 * n
```
Comment (TurtlePU) [inline tests/Tests/UInt.hs:69]: Looks like a constraint we can get rid of

File: src/ZkFold/Symbolic/Algorithms/Hash/SHA2.hs [hunk 14]
```diff
@@ -175,8 +177,9 @@ type family PaddedLength (msg :: Natural) (block :: Natural) (lenBits :: Natural
 type SHA2 algorithm context k =
    ( AlgorithmSetup algorithm context
    , KnownNat k
+   , KnownNat (ChunkSize algorithm)
    , KnownNat (PaddedLength k (ChunkSize algorithm) (2 * WordSize algorithm))
```
Comment (TurtlePU) [inline src/ZkFold/Symbolic/Algorithms/Hash/SHA2.hs:181]: This should be created by us inside `sha2`

File: src/ZkFold/Symbolic/Algorithms/Hash/SHA2.hs [hunk 15]
```diff
@@ -175,8 +177,9 @@ type family PaddedLength (msg :: Natural) (block :: Natural) (lenBits :: Natural
...
-   , (Div (PaddedLength k (ChunkSize algorithm) (2 * WordSize algorithm)) (ChunkSize algorithm)) * (ChunkSize algorithm) ~ PaddedLength k (ChunkSize algorithm) (2 * WordSize algorithm)
+   , Div (PaddedLength k (ChunkSize algorithm) (2 * WordSize algorithm)) (ChunkSize algorithm) * (ChunkSize algorithm) ~ PaddedLength k (ChunkSize algorithm) (2 * WordSize algorithm)
    , k <= PaddedLength k (ChunkSize algorithm) (2 * WordSize algorithm)
```
Comment (TurtlePU) [inline src/ZkFold/Symbolic/Algorithms/Hash/SHA2.hs:183]: This should be created by us inside `sha2`

File: src/ZkFold/Symbolic/Algorithms/Hash/SHA2.hs [hunk 16]
```diff
@@ -175,8 +177,9 @@ type family PaddedLength (msg :: Natural) (block :: Natural) (lenBits :: Natural
 type SHA2 algorithm context k =
    ( AlgorithmSetup algorithm context
    , KnownNat k
+   , KnownNat (ChunkSize algorithm)
```
Comment (TurtlePU) [inline src/ZkFold/Symbolic/Algorithms/Hash/SHA2.hs:180]: Looks like something that should be a part of `AlgorithmSetup`

File: src/ZkFold/Base/Protocol/Plonk/Prover.hs [hunk 17]
```diff
@@ -49,23 +49,23 @@ plonkProve PlonkupProverSetup {..}
         secret i = ps !! (i -! 1)
 
         n = value @n
-        zhX = polyVecZero @_ @n @(PlonkupPolyExtendedLength n)
+        zhX = with4n6 @n $ polyVecZero @_ @n @(PlonkupPolyExtendedLength n)
```
Comment (TurtlePU) [inline src/ZkFold/Base/Protocol/Plonk/Prover.hs:52]: Maybe a single `with4n6` in the beginning of a `plonkProve` would do?

**Review by TurtlePU — CHANGES_REQUESTED** [review]
I think a little more needs to be done, see review comments

File: src/ZkFold/Symbolic/Algorithms/Hash/Blake2b.hs [hunk 18]
```diff
@@ -179,40 +210,25 @@ blake2b key input =
...
     ) => ByteString (8 * inputLen) c -> ByteString 256 c
 blake2b_256 = blake2b @0 @inputLen @32 (fromConstant @Natural 0)
 
 -- | Hash a `ByteString` using the Blake2b-256 hash function.
 blake2b_512 :: forall inputLen c n .
     ( Symbolic c
     , KnownNat inputLen
-    , KnownNat (ExtensionBits inputLen)
-    , KnownNat (8 * inputLen)
     , n ~ (8 * inputLen + ExtensionBits inputLen)
```
Comment (TurtlePU) [inline src/ZkFold/Symbolic/Algorithms/Hash/Blake2b.hs:232]: Is this parameter needed?

File: src/ZkFold/Symbolic/Algorithms/Hash/SHA2.hs [hunk 19]
```diff
@@ -170,13 +174,29 @@ type family NextMultiple (n :: Natural) (divisor :: Natural) :: Natural where
...
+withPaddedLength = withDict (withPaddedLength' @n @d @l)
+
 -- | Constraints required for a type-safe SHA2
 --
 type SHA2 algorithm context k =
    ( AlgorithmSetup algorithm context
    , KnownNat k
-   , KnownNat (PaddedLength k (ChunkSize algorithm) (2 * WordSize algorithm))
-   , (Div (PaddedLength k (ChunkSize algorithm) (2 * WordSize algorithm)) (ChunkSize algorithm)) * (ChunkSize algorithm) ~ PaddedLength k (ChunkSize algorithm) (2 * WordSize algorithm)
+   , Div (PaddedLength k (ChunkSize algorithm) (2 * WordSize algorithm)) (ChunkSize algorithm) * (ChunkSize algorithm) ~ PaddedLength k (ChunkSize algorithm) (2 * WordSize algorithm)
```
Comment (TurtlePU) [inline src/ZkFold/Symbolic/Algorithms/Hash/SHA2.hs:199]: This should be created by us inside `sha2`

File: src/ZkFold/Symbolic/Data/ByteString.hs [hunk 20]
```diff
@@ -117,35 +118,11 @@ class ShiftBits a where
...
-
--- | Describes types which can be made by concatenating several words of equal length.
---
-class Concat a b where
-    concat :: [a] -> b
-
-
 -- | Describes types that can be truncated by dropping several bits from the end (i.e. stored in the lower registers)
 --
 class Truncate a b where
```
Comment (TurtlePU) [inline src/ZkFold/Symbolic/Data/ByteString.hs:123]: Do we need this as a class?

File: src/ZkFold/Symbolic/Algorithms/Hash/Blake2b.hs [hunk 21]
```diff
@@ -132,37 +135,65 @@ blake2b' d =
...
+
+with8nLessExt :: forall n {r}. KnownNat n => (8 * n <= 8 * n + ExtensionBits n => r) -> r
+with8nLessExt = withExtendedInputByteString @n $
+                    withDict (zeroLe @(ExtensionBits n)) $
+                        withDict (plusMonotone2 @(8 * n) @0 @(ExtensionBits n))
+
+with8n  :: forall n {r}. KnownNat n => (KnownNat (8 * n) => r) -> r
+with8n = withDict (timesNat @8 @n)
+
+black2bDivConstraint :: forall n. Dict (Div (8 * n + ExtensionBits n) 64 * 64 ~ 8 * n + ExtensionBits n)
```
Comment (vlasin) [inline src/ZkFold/Symbolic/Algorithms/Hash/Blake2b.hs:163]: It's `blake2b`

File: src/ZkFold/Symbolic/Algorithms/Hash/Blake2b.hs [hunk 22]
```diff
@@ -132,37 +135,65 @@ blake2b' d =
...
+black2bDivConstraint = unsafeAxiom
+
+withBlack2bDivConstraint :: forall n {r}. (Div (8 * n + ExtensionBits n) 64 * 64 ~ 8 * n + ExtensionBits n => r) -> r
```
Comment (vlasin) [inline src/ZkFold/Symbolic/Algorithms/Hash/Blake2b.hs:166]: It's `blake2b`

File: examples/Examples/ByteString.hs [hunk 23]
```diff
@@ -34,5 +37,9 @@ exampleByteStringAdd ::
   forall n c. (KnownNat n, Symbolic c) => ByteString n c -> ByteString n c -> ByteString n c
 exampleByteStringAdd x y = from (from x + from y :: UInt n Auto c)
 
-exampleSHA :: forall n c. SHA2 "SHA256" c n => ByteString n c -> ByteString 256 c
+exampleSHA :: forall n c.
+  SHA2 "SHA256" c n
+  => Div (PaddedLength n 512 64) 512 * 512 ~ PaddedLength n 512 64
```
Comment (TurtlePU) [inline examples/Examples/ByteString.hs:42]: We need to get rid of these constraints by cooking up the evidence inside `sha2` where it's needed

File: examples/Examples/ByteString.hs [hunk 24]
```diff
@@ -34,5 +37,9 @@ exampleByteStringAdd ::
...
+  => n <= PaddedLength n 512 64
```
Comment (TurtlePU) [inline examples/Examples/ByteString.hs:43]: This is always true by the definition of `PaddedLength` and should be cooked up inside `sha2`

File: src/ZkFold/Base/Data/Vector.hs [hunk 25]
```diff
@@ -175,3 +176,7 @@ instance (Random a, KnownNat size) => Random (Vector size a) where
 
 instance ToJSON a => ToJSON (Vector n a) where
     toJSON (Vector xs) = toJSON xs
+
+
+parFmap :: (a -> b) -> Vector size a -> Vector size b
```
Comment (TurtlePU) [inline src/ZkFold/Base/Data/Vector.hs:181]: It is unclear whether this would provide any benefits (also considering conversion to list and back), so I'd advice to just use `<$>`. In case we need parallelism, using one from `Data.Vector.Strategies` is better, I guess

File: src/ZkFold/Symbolic/Algorithms/Hash/Blake2b.hs [hunk 26]
```diff
@@ -132,37 +135,65 @@ blake2b' d =
...
+
+blake2b :: forall keyLen inputLen outputLen c n.
     ( Symbolic c
     , KnownNat keyLen
     , KnownNat inputLen
     , KnownNat outputLen
-    , KnownNat (ExtensionBits inputLen)
-    , KnownNat (8 * inputLen)
-    , KnownNat (8 * outputLen)
     , n ~ (8 * inputLen + ExtensionBits inputLen)
```
Comment (TurtlePU) [inline src/ZkFold/Symbolic/Algorithms/Hash/Blake2b.hs:183]: Here `n` also looks unused

File: src/ZkFold/Symbolic/Algorithms/Hash/SHA2.hs [hunk 27]
```diff
@@ -186,56 +204,66 @@ type SHA2 algorithm context k =
...
+divisibleDiv :: forall a b. (Mod a b ~ 0) :- (Div a b * b ~ a)
+divisibleDiv = Sub unsafeAxiom
+
+withDivisibleDiv :: forall a b {r}. Mod a b ~ 0 => ((Div a b * b ~ a) => r) -> r
+withDivisibleDiv = withDict (divisibleDiv @a @b)
+
 sha2
     :: forall (algorithm :: Symbol) context k
     .  SHA2 algorithm context k
+    => Div (PaddedLength k (ChunkSize algorithm) (2 * WordSize algorithm)) (ChunkSize algorithm) * (ChunkSize algorithm) ~ PaddedLength k (ChunkSize algorithm) (2 * WordSize algorithm)
```
Comment (TurtlePU) [inline src/ZkFold/Symbolic/Algorithms/Hash/SHA2.hs:217]: Evidence for these should be cooked up inside `sha2`, as these are semantically true and should not be present in the public interface

File: src/ZkFold/Symbolic/Data/ByteString.hs [hunk 28]
```diff
@@ -232,31 +207,26 @@ instance (Symbolic c, KnownNat n) => BoolType (ByteString n c) where
...
-instance
-  ( Symbolic c
-  , KnownNat wordSize
-  , (Div n wordSize) * wordSize ~ n
-  ) => ToWords (ByteString n c) (ByteString wordSize c) where
-    toWords (ByteString bits) = Haskell.map (ByteString . packed) $ V.fromVector . V.chunks @(Div n wordSize) @wordSize $ unpacked bits
+toWords :: forall m wordSize c. (Symbolic c, KnownNat wordSize) => ByteString (m * wordSize) c -> Vector m (ByteString wordSize c)
+toWords (ByteString bits) = parFmap ByteString $ unpackWith (V.chunks @m @wordSize) bits
 
 -- | Unfortunately, Haskell does not support dependent types yet,
```
Comment (TurtlePU) [inline src/ZkFold/Symbolic/Data/ByteString.hs:213]: This comment can be removed, as `concat` is completely sound

File: symbolic-base/src/ZkFold/Base/Data/Vector.hs [hunk 29]
```diff
@@ -6,6 +6,7 @@ module ZkFold.Base.Data.Vector where
 
 import           Control.DeepSeq                  (NFData)
 import           Control.Monad.State.Strict       (runState, state)
+import           Control.Parallel.Strategies      (parMap, rpar)
```
Comment (TurtlePU) [inline symbolic-base/src/ZkFold/Base/Data/Vector.hs:9]: Unused import

**Review by TurtlePU — APPROVED** [review]
Nice! 
I think we can merge after checking that tests pass

**Comment by hovanja2011** [comment]
All tests passed
